File: packages/server-python/app/schemas/milito.py
"""Milito game schemas."""

from typing import Optional

from pydantic import BaseModel


class MilitoPlayerTableResponse(BaseModel):
    """Response schema for player's battlefield table."""

    enemy_row_2: list[Optional[int]]
    enemy_row_1: list[Optional[int]]
    territory_row: list[int]
    player_row_1: list[Optional[int]]
    player_row_2: list[Optional[int]]


class MilitoPlayerStateResponse(BaseModel):
    """Response schema for a player's state in Milito."""

    oderId: str
    faction: str
    table: MilitoPlayerTableResponse
    hand: list[int]
    phase: str
    cardsToDiscard: int
    selectedCard: Optional[int] = None
    selectedColumn: Optional[int] = None
    discardedCards: list[int]
    score: int


class MilitoGameStateResponse(BaseModel):
    """Response schema for Milito game state."""

    players: dict[str, MilitoPlayerStateResponse]
    currentPlayerId: str
    turnNumber: int
    status: str
    winnerId: Optional[str] = None


class MilitoStateResponse(BaseModel):
    """Response schema for complete Milito state."""

    room: dict
    gameState: MilitoGameStateResponse
    myState: MilitoPlayerStateResponse
    players: list[dict]
    isMyTurn: bool


class SelectCardRequest(BaseModel):
    """Request schema for selecting a card."""

    cardIndex: int


class SelectColumnRequest(BaseModel):
    """Request schema for selecting a column."""

    columnIndex: int


class DiscardRequest(BaseModel):
    """Request schema for discarding a card."""

    cardIndex: int

File: src/shared/types/milito.ts
// Milito Game Types

export type MilitoFaction = 'alexandrian_macedonian' | 'ancient_british'

export const MILITO_UNIT_TYPES = [
  'light_cavalry',
  'slingers',
  'leader_2',
  'chariots',
  'heavy_cavalry',
] as const

export type MilitoUnitType = (typeof MILITO_UNIT_TYPES)[number]

export interface MilitoCard {
  id?: number
  unitType: MilitoUnitType
}

export interface MilitoPlayerTable {
  enemy_row_2: (number | null)[]
  enemy_row_1: (number | null)[]
  territory_row: number[]
  player_row_1: (number | null)[]
  player_row_2: (number | null)[]
}

// Game state phases
export const MILITO_SELECT_CARD = 'SELECT_CARD'
export const MILITO_SELECT_COLUMN = 'SELECT_COLUMN'
export const MILITO_DISCARD = 'DISCARD'

export type MilitoPhase =
  | typeof MILITO_SELECT_CARD
  | typeof MILITO_SELECT_COLUMN
  | typeof MILITO_DISCARD

export interface MilitoPlayerState {
  oderId: string
  faction: MilitoFaction
  table: MilitoPlayerTable
  hand: number[]
  phase: MilitoPhase
  cardsToDiscard: number
  selectedCard?: number
  selectedColumn?: number
  discardedCards: number[]
  score: number
}

export interface MilitoGameState {
  players: Record<string, MilitoPlayerState>
  currentPlayerId: string
  turnNumber: number
  status: 'playing' | 'finished'
  winnerId?: string
}
